### ISSUE TITLE
Data Providers

### ISSUE THREAD

**Issue by pottedmeat** [dojo/dgrid#4]
## Overview

As part of dgrid development, we've identified data providers - objects that can be instructed of state changes and notify observers as a result - as being a useful reactive pattern to drive grid interaction.

We should identify what we expect to encounter to ensure there are no interactions that could become problematic as the grid continues to expand in functionality:

* An interface with the methods to call on the data provider to notify it of state changes
* A state object that tracks the grid's initial configuration plus any state changes
* What, if any, state changes should result in other states being changed
* Information that clarifies the capabilities and state of the returned items within the larger data set
* Information that clarifies the capabilities and state of each item within the returned data

## Data Provider Internal State

This data is available to all data provider implementations.

* `size: { dataLength: number; totalLength: number; }`
* `sort: { columnId: string; direction: 'asc' | 'desc' }[]`
* `slice: { start: number; count: number; }`
* `limit: { start: number; count: number; }`
* `expanded: { [ itemId: string ]: boolean }`
* `selected: { [ itemId: string ]: string[] }`
* Pagination would likely add further information to the state, though many implementations could choose to ignore it:
	* `page: { page: number; itemsPerPage: number; }`
* We should consider an object to represent `place(...)` operations (see Callbacks below) to avoid mutating the data directly for DnD operations:
	* `place: Array<{ items: ItemProperties[], position: string | number, relativeTo?: ItemProperties }>`

## Item Metadata

Each item within the data is "marked up" by the data provider with information about each item. This allows configuration concerning data structure to be implemented by the data provider, not by the grid. For example, `idProperty` can be passed to the data provider which uses that to assign the `id` to its properties.

* `id: string`
* `index: number`
* `expandedLevel?: number` indicates how many "parents" were expanded to reach this item
* `isExpanded?: boolean`
* `canExpand?: boolean`
* `isSelected?: enum.ROW | string[]` either an enum (whole row) or a list of column IDs
* `data: T` (the item data, generically typed)

## Data Metadata

This is the object passed to observers. As part of this metadata, the item metadata will be passed in a flat array.

* `size: { dataLength: number; totalLength: number; }`
	* `dataLength` providers information about how many items there are in the subset created by `limit` state changes regardless of how many are in the items array
	* `totalLength` provides information about how many items there are in total regardless of how many are in the items array
* `sort: { columnId: string; direction: 'asc' | 'desc' }[]`
* `limit: { start: number; count: number; }`
	* Explains that the grid will be traversing a subset of the overall data and that slices will happen from this data set
* `slice: { start: number; count: number; }`
	* Explains that this is only a subset of the overall data the grid can traverse
* `items: ItemProperties<T>[]`

## Callbacks

These are the methods we expect the grid (or other consumers) to call:

* `sort(sorting: SortDetails[]): void;`
* `slice({ start: number, count: number }): void;`
* `limit({ start: number, count: number }): void;`
* `toggleExpanded(item: ItemProperties): void`
* `toggleSelected(item: ItemProperties, columnId?: column }`
* `place(items: ItemProperties[], position: string | number, relativeTo?: ItemProperties)` (for drag-and-drop of rows where position may be before, after, index, etc.)
* Pagination will call a "paging" method with a default implementation in the base data provider that calls `slice` but can be overridden by an implementation if they choose to deal with paging differently:
	* `slicePage: { page: number; itemsPerPage: number; }: void;`
* `configure(...): void` (would accept all state config in a single call)
* `expand(item: ItemProperties): void`
* `collapse(item: ItemProperties): void`

**Comment by dylans** [comment]
`place` strikes me as a bit odd though I'm not sure I have a more obvious name (move or position, though you then have position as a parameter, so I guess it's a fine name).

I assume expandedLevel represents how many levels something is open to? This is per item or per hierarchy or ?

For size, please clarify the distinction beween size and totalLength

"Because of this" stray thought at the end?

**Comment by pottedmeat** [comment]
The reason I ended up with a weird name like `place` is that I was trying to let it do double-duty for both moving and for adding content - like `dojo/place` does.

There is no hierarchy in the data, so `expandedLevel` would be per-item and represents how many parents have been opened to reach that item.

**Comment by dylans** [comment]
> The reason I ended up with a weird name like place is that I was trying to let it do double-duty for both moving and for adding content - like dojo/place does.

yeah, I can live with it unless someone has a better suggestion. :)

> There is no hierarchy in the data, so expandedLevel would be per-item and represents how many parents have been opened to reach that item.

ok, makes sense, maybe explain that above?

**Comment by pottedmeat** [comment]
Updated

**Comment by pottedmeat** [comment]
Added a note on typing as my current plan to use generics/class mixins looks like it may not work.

**Comment by kitsonk** [comment]
Looking at this, I have some questions about the fundamentals which are alluding me.  I think the concept with the data providers, from a data consumption side, need to meet the following use cases:

- As a data consumer, I need to understand the capabilities of a data provider:
  - Can the data provider offer me observation/read capability?
  - Can the data provider offer me update capability?
  - Can the data provider offer me create capability?
  - Can the data provider offer me delete capability?
  - Can the data provider offer me transactional capabilities?
- As a data consumer, with a data provider that has observation/read capability, I need to ability to specify what data I need.  I should be able to specify this in the following way:
  - An abstract concept of a query which constrains the overall data set I will want to obtain.
  - An abstract concept of how I want that data shaped.
  - An abstract concept of how I want that data ordered.
- Once I have specified the query, shape and order of the data, I will also want to specify my current view of that data I would like to observe.
  - There should be an API to get meta data about the data set, after it has been declared.  That could be the number of total objects in the data set, a number of objects in the current view, an offset of the view in the total data set.
- I then want that data to be delivered to me, observing any changes, until I terminate my subscription.
- I will need a way to update the parameters of my current view, until I terminate my subscription, but if I wish to change the query, shape or order of the data, I will terminate my subscription and request a new one.
- With a data provider that has update capability, any actor who has a reference to a data object can request that that object be updated, providing the either a replacement object or a partial object containing the changes.
  - The actor requesting the update will be notified the update is resolved or rejected.  Resolution though would not provide any details about what the current value of the object is and the actor should make no assumptions.
  - Any data consumers with this object in their current view will get notified of the update via their subscription.
- With a data provider that has create capability, any actor can request to add a data object, providing as much of the object as it is aware of, with the data provider handling the details of the creation.  The actor would only be notified of the update is resolved or rejected.
- With a data provider that has delete capability, any actor can request to delete a data object.  The actor would only be notified that the delete is resolved or rejected.  Subscribers who have that record in their current view would be notified of the delete.
- Transactional capabilities would _wrap_ create/update/delete operations into a single transaction which would be resolved or rejected when committed.

I really think we should avoid any concepts of internal implementation and focus on contracting the external APIs.  The reason is that there maybe implementations where it could be better to offload some of the internal concepts upstream.  For example, many databases/object repos have concepts of _views_ which would be able to express some of the sort capabilities.  It would be challenging to assume those will always be implemented internally.
